pkgs/servers/matrix-synapse/default.nix

```diff
@@ -44,47 +46,93 @@ buildPythonApplication rec {
     rustc
   ]);
 
-  buildInputs = [ openssl ];
+  buildInputs = [
+    openssl
+  ];
 
-  propagatedBuildInputs = [
-    authlib
+  propagatedBuildInputs = with python3.pkgs; [
+    attrs
     bcrypt
     bleach
     canonicaljson
-    daemonize
+    cryptography
     frozendict
     ijson
     jinja2
     jsonschema
-    lxml
     matrix-common
     msgpack
     netaddr
+    packaging
     phonenumbers
     pillow
     prometheus-client
-    psutil
-    psycopg2
     pyasn1
+    pyasn1-modules
     pydantic
-    pyicu
     pymacaroons
-    pynacl
     pyopenssl
-    pysaml2
     pyyaml
-    requests
-    setuptools
+    service-identity
     signedjson
     sortedcontainers
     treq
     twisted
     typing-extensions
     unpaddedbase64
-  ] ++ lib.optional enableSystemd systemd
-    ++ lib.optionals enableRedis [ hiredis txredisapi ];
+  ]
+  # extras
+  ++ twisted.optional-dependencies.tls;
 
-  nativeCheckInputs = [ mock parameterized openssl ];
+  passthru.optional-dependencies = with python3.pkgs; {
```

Comment:
One way we coulld this more pleasnt would be to write out the environment from the module (`environment.etc."matrix-synapse/environment" = {}`) and source that in the systemd unit. That way at least it becomes reusable.

Reply:
>Allowing users to rely on the cached package is inherently incompatible with including a wrapper for optional dependencies

1. The module could use the non-wrapped version.
1. Building a wrapper is very common and should not be compared with building the whole package

>In fact the module already relies on the PYTHONPATH for installing plugins, that is only provided through the module.

There are barely any plugins and they are not _core_ functionality

>Lastly I don't think running synapse from a terminal window is a valid use case except for debugging. In which case replicating the PYTHONPATH from the systemd unit would be a straight forward way to go.

There are other reasons to want have working binaries without having to set env variables manually. https://nixery.dev/ for example. Or to be compatible with nix sbom tooling.


Having a standard that applications should work, and their runtime dependencies should be managed and represented by the actual application closure is one of the main selling points of nix for me.

Reply:
> There are other reasons to want have working binaries without having to set env variables manually. nixery.dev for example. Or to be compatible with nix sbom tooling.

I would argue that applicatons that evaluate at the sbom should also take a look at the runtime environment as specified in the systemd unit, because lots of things can be injected there.

Not sure what the use case for ad-hoc docker containers would be. So then I have `matrix-synapse` in docker, what now?

Coming back to the original question of the thread:

> This is pretty cool, but going from ${pkgs.matrix-synapse}/bin/synapse just working to needing some wrapper script to set PYTHONPATH is a negative. Could you showcase how one would use this package without NixOS with this change?

By overwriting the package and extending `propagatedBuildInputs`. The negative here is that you will have to build the package from source and run the test suite.

```
nix-repl> :b matrix-synapse.overridePythonAttrs (oldAttrs: { propagatedBuildInputs = oldAttrs.propagatedBuildInputs ++ oldAttrs.passthru.optional-dependencies.oidc; })
```

> Maybe pkgs.matrix-synapse could be a functioning matrix-synapse (by wrapping or some other way), and then moving all this stuff into python3.pkgs.matrix-synapse could let us keep nixpkgs' way of neatly enclosing runtime dependencies while still keeping this?

The `python3.pkgs` package set is only intended for libraries, not applications. The relevant part here is, that only applications can use `packageOverrides` to match required dependency versions.

The only way to make the package work in the intended scenario is to modify the `propagatedBuildInputs` as demonstrated above, and generate a package that includes the necessary dependencies.

Reply:
Personally I don't want to invest time into creating a wrapper function. I don't think they're worth the effort, because they are inherently undiscoverable and require separate documentation, that nobody will know where to look for.

The same goes for the `enableFoo` arguments, which I'm dropping in this changeset.

Reply:
Just in case it wasn't clear, I agree about the `enable*`-arguments. I wanted to point out that I don't think that a big wrapper containing a bunch of optional things is appropriate here, i.e. I'm in favor of your approach.

Reply:
>I would argue that applicatons that evaluate at the sbom should also take a look at the runtime environment as specified in the systemd unit, because lots of things can be injected there.

>Not sure what the use case for ad-hoc docker containers would be. So then I have matrix-synapse in docker, what now?

I am sure you understood that my examples were just that: examples of external tooling that make (fairly reasonable) assumptions. And as a counter to the strawman that "you shouldn't run synapse in a terminal anyway". And thus arguing their specific merits are beside the point.

As for the sboms: making a an sbom of a package and an sbom for a system configuration are two different things. Having to evaulate a nixos module, and extract the (collection of) relevant systemd units vs just giving it a nixpkgs drv are two entirely different things and arguing that the latter is useless is just bad faith.

>The python3.pkgs package set is only intended for libraries, not applications. The relevant part here is, that only applications can use packageOverrides to match required dependency versions.
>The only way to make the package work in the intended scenario is to modify the propagatedBuildInputs as demonstrated above, and generate a package that includes the necessary dependencies. 

Why is it when the wrapping is done in systemd it's fine to put it in PYTHONPATH but when doing it in a shell script that approach suddenly doesn't work? Am I missing something important here?

>That may be true for plain nixpkgs use, but hard disagree for NixOS use. The runtime environment will often make or break an application.

We're discussing a package in nixpkgs, and how to control its runtime environment/dependencies. Just because this package also has a NixOS service does not mean we should just hose all other consumers.

> How do you define core functionality?

`postgres`, `systemd`, `url-preview`, `redis` (workers), `cache-memory`, `user-search`

Im unsure about `jwt` I think that might(!) be involved in some widget auth stuff, but I'm not sure how.

In the (near<sup>matrix time</sup>) future also `oidc` as that will be the only way to autheticate with synapse.

>Where would you draw the line between plugins that are supposed to be part of said package and which aren't? Also, enable* flags to toggle some behavior (especially things you don't want to ship by default) are hard to discover.

I'd be in favor of dropping the `enableSystemd` and `enableRedis` flags.

As for the wrapper package I don't really care how it's done as long as it has reasonable defaults (like the set above).

If the rest of the dependencies are only a few megabytes I'd probably just add them as well to simplify the package.

There's exactly a thriving ecosystem of plugins so it's not really a problem.

[modules](https://matrix-org.github.io/synapse/latest/modules/index.html) however would also need to be considered...

>looking at the changes here this should still be the case

Only in the NixOS service. The package `matrix-synapse` would have no reason to keep existing.

You could just move the entire package to a let block at top of the module and it'd be more accurate. (well, other than breaking overlays).

>Also, enable* flags to toggle some behavior (especially things you don't want to ship by default) are hard to discover

They are, but I'd argue this incantation:
`pkgs.matrix-synapse.overridePythonAttrs (oldAttrs: { propagatedBuildInputs = oldAttrs.propagatedBuildInputs ++ (with oldAttrs.passthru.optional-dependencies; [ postgres systemd url-preview user-search redis cache-memory oidc ]); })"`
is even harder. (Wtf it can flatten a list of lists?)

Reply:
> systemd

synapse seems supported on platforms w/o systemd support (see discussion above)

> redis

We don't even have worker support in nixpkgs and even if it will be at some point, it won't be the default.

> They are, but I'd argue this incantation:

That shouldn't be needed at all, the optional dependencies only need to be available in PYTHONPATH, also that'll save you a rebuild.

> [modules](https://matrix-org.github.io/synapse/latest/modules/index.html) however would also need to be considered...

Correct, but I think that this is a good first step in the right direction :) 

Personally, I think an option would be to move the `PYTHONPATH = ...` stuff into a wrapper function in `nixpkgs` then it can be re-used in alternative modules (IIRC your initial remark). Defaults should IMHO remain the one @mweinelt proposed, but rebuilds will be cheap then so I don't consider this a big deal.

Reply:
> > systemd
> 
> synapse seems supported on platforms w/o systemd support (see discussion above)

You'd select it by default on platforms which support it

> > redis
> 
> We don't even have worker support in nixpkgs and even if it will be at some point, it won't be the default.

This will change soon, and I'd argue some workers are actually reasonable to set up by default in an opinionated setup. But that is a discussion for another day.

> Defaults should IMHO remain the one @mweinelt proposed, but rebuilds will be cheap then so I don't consider this a big deal.

Alright, but at that point we'd be arguing over a closure size reduction of <30MB with the added risks of upstream not testing the full matrix (heh) and a more complex package.

> Personally, I think an option would be to move the PYTHONPATH = ... stuff into a wrapper function in nixpkgs then it can be re-used in alternative modules (IIRC your initial remark). Defaults should IMHO remain the one @mweinelt proposed, but rebuilds will be cheap then so I don't consider this a big deal.

Right, so I see this as essentially:

```nix
matrix-synapse = { matrix-synapse-unwrapped, extras ? [ "postgres" (if available then "systemd" else null) "url-preview" "user-search" ], ...}: let
  matrix-synapse-unwrapped = { ... }:  python3.pkgs.buildPythonApplication {...};
in stdenv.mkDerivation {
  nativeBuildInputs = [ makeWrapper ];
  ...
  buildPhase = ''
  	for b in ${matrix-synapse-unwrapped}/bin/*; do
		makeWrapper \
			--set PYTHONPATH=${makeSearchPathOutput "lib" cfg.package.python.sitePackages [ extrasEnv ];}
			"$b" "$out/bin"
	done;
  '';
  passthru.unwrapped = matrix-synapse-unwrapped;
};
```

Then the module can override the extras without needing loss of functionality for non-nixos

Reply:
Looks like makeWrapper isn't the right tool for the job. Probably needs `wrapPython`? Not sure.

```
matrix-synapse-wrapped> building '/nix/store/2528iqizypzbizvqkg70xzgnfwibw462-matrix-synapse-wrapped-1.84.1.drv'
matrix-synapse-wrapped> -r-xr-xr-x 2 nobody nogroup 3.3K Jan  1  1970 /nix/store/gnk2pmcpwwnc4kb6634pmky1ryf16i64-matrix-synapse-1.84.1/bin/export_signing_key
matrix-synapse-wrapped> -r-xr-xr-x 2 nobody nogroup 3.3K Jan  1  1970 /nix/store/gnk2pmcpwwnc4kb6634pmky1ryf16i64-matrix-synapse-1.84.1/bin/generate_config
matrix-synapse-wrapped> -r-xr-xr-x 2 nobody nogroup 3.3K Jan  1  1970 /nix/store/gnk2pmcpwwnc4kb6634pmky1ryf16i64-matrix-synapse-1.84.1/bin/generate_log_config
matrix-synapse-wrapped> -r-xr-xr-x 2 nobody nogroup 3.3K Jan  1  1970 /nix/store/gnk2pmcpwwnc4kb6634pmky1ryf16i64-matrix-synapse-1.84.1/bin/generate_signing_key
matrix-synapse-wrapped> -r-xr-xr-x 2 nobody nogroup 3.3K Jan  1  1970 /nix/store/gnk2pmcpwwnc4kb6634pmky1ryf16i64-matrix-synapse-1.84.1/bin/hash_password
matrix-synapse-wrapped> -r-xr-xr-x 2 nobody nogroup 3.3K Jan  1  1970 /nix/store/gnk2pmcpwwnc4kb6634pmky1ryf16i64-matrix-synapse-1.84.1/bin/register_new_matrix_user
matrix-synapse-wrapped> -r-xr-xr-x 2 nobody nogroup 3.3K Jan  1  1970 /nix/store/gnk2pmcpwwnc4kb6634pmky1ryf16i64-matrix-synapse-1.84.1/bin/synapse_homeserver
matrix-synapse-wrapped> -r-xr-xr-x 2 nobody nogroup 3.3K Jan  1  1970 /nix/store/gnk2pmcpwwnc4kb6634pmky1ryf16i64-matrix-synapse-1.84.1/bin/synapse_port_db
matrix-synapse-wrapped> -r-xr-xr-x 2 nobody nogroup 3.3K Jan  1  1970 /nix/store/gnk2pmcpwwnc4kb6634pmky1ryf16i64-matrix-synapse-1.84.1/bin/synapse_review_recent_signups
matrix-synapse-wrapped> -r-xr-xr-x 2 nobody nogroup 3.3K Jan  1  1970 /nix/store/gnk2pmcpwwnc4kb6634pmky1ryf16i64-matrix-synapse-1.84.1/bin/synapse_worker
matrix-synapse-wrapped> -r-xr-xr-x 2 nobody nogroup 3.3K Jan  1  1970 /nix/store/gnk2pmcpwwnc4kb6634pmky1ryf16i64-matrix-synapse-1.84.1/bin/synctl
matrix-synapse-wrapped> -r-xr-xr-x 2 nobody nogroup 3.3K Jan  1  1970 /nix/store/gnk2pmcpwwnc4kb6634pmky1ryf16i64-matrix-synapse-1.84.1/bin/update_synapse_database
matrix-synapse-wrapped> 
matrix-synapse-wrapped> Builder called die: Cannot wrap '--set' because it is not an executable file
matrix-synapse-wrapped> Backtrace:
matrix-synapse-wrapped> 7 assertExecutable /nix/store/2y2nabl483f2ri15x11b4sv03ds6z15x-make-shell-wrapper-hook/nix-support/setup-hook
matrix-synapse-wrapped> 43 makeShellWrapper /nix/store/2y2nabl483f2ri15x11b4sv03ds6z15x-make-shell-wrapper-hook/nix-support/setup-hook
matrix-synapse-wrapped> 36 makeWrapper /nix/store/2y2nabl483f2ri15x11b4sv03ds6z15x-make-shell-wrapper-hook/nix-support/setup-hook
matrix-synapse-wrapped> 1552 genericBuild /nix/store/0swa08bbslxc9wbfixbcd196aa1117lx-stdenv-linux/setup
matrix-synapse-wrapped> 6 main /nix/store/6xg259477c90a229xwmb53pdfkn6ig3g-default-builder.sh
matrix-synapse-wrapped>

Reply:
Okay, looks fixed now. PTAL.